subsys/bluetooth/controller/util/mfifo.h

```diff
@@ -0,0 +1,177 @@
+
+#define MFIFO_DEFINE(name, sz, cnt) \
+		struct { \
+			u8_t const s; /* TODO: const, optimise RAM use */ \
+			u8_t const n; /* TODO: const, optimise RAM use */ \
+			u8_t f; \
+			u8_t l; \
+			u8_t m[sz * ((cnt) + 1)]; \
+		} mfifo_##name = { \
+			.n = ((cnt) + 1), \
+			.s = (sz), \
+			.f = 0, \
+			.l = 0, \
+		}
+
+#define MFIFO_INIT(name) \
+	mfifo_##name.f = mfifo_##name.l = 0
+
+static inline bool mfifo_enqueue_idx_get(u8_t count, u8_t first, u8_t last,
+					 u8_t *idx)
+{
+	last = last + 1;
+	if (last == count) {
+		last = 0;
+	}
+
+	if (last == first) {
+		return false;
+	}
+
+	*idx = last;
+
+	return true;
+}
+
+#define MFIFO_ENQUEUE_IDX_GET(name, i) \
+		mfifo_enqueue_idx_get(mfifo_##name.n, mfifo_##name.f, \
+				      mfifo_##name.l, (i))
+
+static inline void mfifo_by_idx_enqueue(u8_t *fifo, u8_t size, u8_t idx,
+					void *mem, u8_t *last)
+{
+	void **p = (void **)(fifo + (*last) * size);
+
+	*p = mem;
```

Comment:
If you need to support non-pointer mfifos then this should be a `memcpy`

Reply:
Agree.